---
title: "R Notebook"
output: html_notebook
---



```{r}
library(data.table)
library(tidyverse)
library(biomaRt)
load("../data/varDat.hg19.RData")
# gene.Names <- dat[, "Gene.refGene"]
# gene.Names <- unique(dat[, "Gene.refGene"])

gene.Names <- fread("uniquegenes.txt")
gene.Names <- unique(gene.Names)

ref <- fread("refFlat.txt", header = F)

# schema from: https://genome.ucsc.edu/goldenpath/gbdDescriptions.html
colnames(ref) <- c("geneName", "name", "chrom", "strand", "txStart", "txEnd", "cdsStart", "cdsEnd", "exonCount", "exonStarts", "exonEnds")
colnames(gene.Names) <- "geneName"

ref$bpdif <- ref$txEnd - ref$txStart

ref <- ref[order(geneName, -bpdif), ]

gene.Names.Anno <- merge(ref, gene.Names, by = "geneName") # [, c("geneName", "name", "strand", "chrom", "txStart", "txEnd")]
genes.to.remove <- gene.Names.Anno[duplicated(gene.Names.Anno$geneName)][,c("geneName", "name", "bpdif")]

gene.Names.Anno.Filtered <- as.data.table(anti_join(gene.Names.Anno, genes.to.remove, by = "name")[, c("geneName", "chrom", "txStart", "txEnd")])

unavailable <- anti_join(gene.Names, gene.Names.Anno.Filtered[, "geneName"])

# gene.Names.Anno <- gene.Names.Anno[order(geneName, bpdif), ]


# gene.Names.Anno.Unique <- unique(gene.Names.Anno)



# 
human <- useMart("ENSEMBL_MART_ENSEMBL", host = "grch37.ensembl.org", dataset = "hsapiens_gene_ensembl")
# 
gene.Names.Anno2 <- as.data.table(getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'), 
                         filters = 'hgnc_symbol',
                         values = unavailable,
                         mart = human)
)

gene.Names.Anno2$bpdif <- gene.Names.Anno2$end_position - gene.Names.Anno2$start_position
gene.Names.Anno2 <- gene.Names.Anno2[order(hgnc_symbol, bpdif), ]
genes.to.remove <- gene.Names.Anno2[duplicated(gene.Names.Anno2$hgnc_symbol)]

gene.Names.Anno2.Filtered <- as.data.table(anti_join(gene.Names.Anno2, genes.to.remove, by = "hgnc_symbol"))[, -c("bpdif")]

# fixing a chromosome position
gene.Names.Anno2.Filtered[13, 2] <- 6

colnames(gene.Names.Anno2.Filtered) <- colnames(gene.Names.Anno.Filtered)

unavailable2 <- anti_join(unavailable, gene.Names.Anno2.Filtered[, "geneName"])


# 
# gene.Names.Anno2 <- as.data.table(getBM(attributes = c('external_gene_name', 'chromosome_name', 'start_position', 'end_position'), 
#                          filters = 'external_gene_name', 
#                          values = gene.Names, 
#                          mart = human)
# )
# 
# colnames(gene.Names.Anno)[1] <- "Gene.refGene"
# colnames(gene.Names.Anno2)[1] <- "Gene.refGene"
# 
# # colnames(gene.Names) <- "Gene.refGene"
# 
# unavailable <- anti_join(gene.Names, gene.Names.Anno[, "Gene.refGene"])
# 
# #colnames(gene.Names) <- "external_gene_name"
# unavailable2 <- anti_join(gene.Names, gene.Names.Anno2[, "Gene.refGene"])
# 
# anti_join(unavailable, unavailable2, by = )
# 
# gene.Names.Anno.numChr <- gene.Names.Anno[grepl("\\d+", hgnc_symbol)]

refGene <- fread("hg19.refGene.txt.gz")[,c(2,3,5,6,13)]
colnames(refGene) <- c("name", "chrom", "txStart", "txEnd", "geneName")

refGene$bpdif <- refGene$txEnd - refGene$txStart

refGene <- refGene[order(geneName, -bpdif), ]
refGene$identifier <- paste0(refGene$geneName, refGene$txStart, refGene$txEnd, refGene$name)


genes.to.remove <- refGene[duplicated(refGene$geneName)][,c("geneName", "name", "bpdif", "identifier")]


refGene.Filtered <- as.data.table(anti_join(refGene, genes.to.remove, by = "identifier")[, c("geneName", "chrom", "txStart", "txEnd")])

gene.Names.Anno3 <- merge(refGene.Filtered, unavailable2, by = "geneName") # [, c("geneName", "name", "strand", "chrom", "txStart", "txEnd")]
# genes.to.remove <- gene.Names.Anno3[duplicated(gene.Names.Anno3$geneName)][,c("geneName", "name", "bpdif")]

#gene.Names.Anno3.Filtered <- as.data.table(anti_join(gene.Names.Anno3, genes.to.remove, by = "name")[, c("geneName", "chrom", "txStart", "txEnd")])

unavailable3 <- anti_join(unavailable2, gene.Names.Anno3[, "geneName"])



gene.Names.Anno.Filtered$chrom <- gsub("chr(\\d+)", "\\1", gene.Names.Anno.Filtered$chrom)
gene.Names.Anno3$chrom <- gsub("chr(\\d+).*", "\\1", gene.Names.Anno3$chrom)

gene.Names.Anno.Aggregate <- rbind(gene.Names.Anno.Filtered, gene.Names.Anno2.Filtered, gene.Names.Anno3)
# include 
gene.Names.Anno.Aggregate$upstream <- NULL
gene.Names.Anno.Aggregate$downstream <- NULL

colnames(gene.Names.Anno.Aggregate) <- c("id", "chr", "37bp1", "37bp2")
gene.Names.Anno.Aggregate$name <- NULL

#gene.Names.Anno.Aggregate <- gene.Names.Anno.Aggregate[,c("id", "chr", "37bp1", "37bp2")]

refname <- fread("hgnc_complete_set.txt")[, c(2,3)]
colnames(refname)[1] <- "id"

#refname2 <- refname %>% rename("refName" = "name")

gene.Names.Anno.Aggregate.Name <- left_join(gene.Names.Anno.Aggregate, refname, by = "id")
gene.Names.Anno.Aggregate.Name$`38bp1` <- ""
gene.Names.Anno.Aggregate.Name$`38bp2` <- ""

gene.Names.Anno.Aggregate.Name <- gene.Names.Anno.Aggregate.Name[ , c("id", "name", "chr", "38bp1", "38bp2", "37bp1", "37bp2")]

# fwrite(gene.Names.Anno.Aggregate.Name, "../www/geneList.txt")
# gene.Names.Anno.Aggregate$upstream <- gene.Names.Anno.Aggregate$txStart - 1000
# gene.Names.Anno.Aggregate$downstream <- gene.Names.Anno.Aggregate$txEnd + 1000

res <- gene.Names.Anno.Aggregate.Name

for (i in 1:nrow(res)) {
      res$id[i] <- paste0('<a id="', res$id[i], '" href="javascript:;" onclick="resClick(this.id)">', res$id[i], '</a>')
}

geneList <- as.data.table(res)

save(geneList, file = "../www/geneList.RData")

#fwrite(res, "../www/geneList.txt")

```

varList processing
```{r}
library(data.table)
library(tidyverse)

load("../data/varDat.hg19.RData")

prevarList <- dat[, c("HG19_ID", "Chr", "Start", "End", "Gene.refGene", "avsnp150")]
colnames(prevarList) <- c("id", "chr", "37bp1", "37bp2", "geneID", "rsID")

for (i in 1:nrow(prevarList)) {
  prevarList$id[i] <- paste0('<a id="',  prevarList$id[i], '" href="javascript:;" onclick="varClick(this.id)">', prevarList$id[i], '</a>')
  #restoreInput(id = paste0("res", i), default = NULL)
}
varList.nested <- list()
# for (i in 1:22) {
#   varList.nested[i] <- prevarList[[grepl(i, prevarList$Chr)]]
# }
varList.nested <- split(prevarList, prevarList$chr)

save(varList.nested, file = "../www/varList.RData")

```

nested variant info page
```{r}
library(data.table)
library(tidyverse)
load("../data/varDat.hg19.RData")

varDat.nested <- list()
varDat.nested <- split(dat, dat$Chr)
varDat.chr1 <- varDat.nested[[1]]
varDat.chr2 <- varDat.nested[[2]]
varDat.chr3 <- varDat.nested[[3]]
varDat.chr4 <- varDat.nested[[4]]
varDat.chr5 <- varDat.nested[[5]]
varDat.chr6 <- varDat.nested[[6]]
varDat.chr7 <- varDat.nested[[7]]
varDat.chr8 <- varDat.nested[[8]]
varDat.chr9 <- varDat.nested[[9]]
varDat.chr10 <- varDat.nested[[10]]
varDat.chr11 <- varDat.nested[[11]]
varDat.chr12 <- varDat.nested[[12]]
varDat.chr13 <- varDat.nested[[13]]
varDat.chr14 <- varDat.nested[[14]]
varDat.chr15 <- varDat.nested[[15]]
varDat.chr16 <- varDat.nested[[16]]
varDat.chr17 <- varDat.nested[[17]]
varDat.chr18 <- varDat.nested[[18]]
varDat.chr19 <- varDat.nested[[19]]
varDat.chr20 <- varDat.nested[[20]]
varDat.chr21 <- varDat.nested[[21]]
varDat.chr22 <- varDat.nested[[22]]

save(varDat.chr1, file = "../varTab/chr1.RData")
save(varDat.chr2, file = "../varTab/chr2.RData")
save(varDat.chr3, file = "../varTab/chr3.RData")
save(varDat.chr4, file = "../varTab/chr4.RData")
save(varDat.chr5, file = "../varTab/chr5.RData")
save(varDat.chr6, file = "../varTab/chr6.RData")
save(varDat.chr7, file = "../varTab/chr7.RData")
save(varDat.chr8, file = "../varTab/chr8.RData")
save(varDat.chr9, file = "../varTab/chr9.RData")
save(varDat.chr10, file = "../varTab/chr10.RData")
save(varDat.chr11, file = "../varTab/chr11.RData")
save(varDat.chr12, file = "../varTab/chr12.RData")
save(varDat.chr13, file = "../varTab/chr13.RData")
save(varDat.chr14, file = "../varTab/chr14.RData")
save(varDat.chr15, file = "../varTab/chr15.RData")
save(varDat.chr16, file = "../varTab/chr16.RData")
save(varDat.chr17, file = "../varTab/chr17.RData")
save(varDat.chr18, file = "../varTab/chr18.RData")
save(varDat.chr19, file = "../varTab/chr19.RData")
save(varDat.chr20, file = "../varTab/chr20.RData")
save(varDat.chr21, file = "../varTab/chr21.RData")
save(varDat.chr22, file = "../varTab/chr22.RData")
```


